components/ephemeral_storage/browser/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,55 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/ephemeral_storage/browser/ephemeral_storage_tab_helper.h"
+
+#include <string>
+
+#include "base/feature_list.h"
+#include "components/content_settings/core/common/features.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/render_view_host.h"
+#include "content/public/browser/web_contents.h"
+#include "third_party/blink/renderer/modules/storage/brave_dom_window_storage.h"
+
+using content::BrowserContext;
+using content::NavigationHandle;
+using content::WebContents;
+
+namespace ephemeral_storage {
+
+EphemeralStorageTabHelper::~EphemeralStorageTabHelper() {}
+
+EphemeralStorageTabHelper::EphemeralStorageTabHelper(WebContents* web_contents)
+    : WebContentsObserver(web_contents) {}
+
+void EphemeralStorageTabHelper::ReadyToCommitNavigation(
+    NavigationHandle* navigation_handle) {
+  if (!navigation_handle->IsInMainFrame())
+    return;
+  if (navigation_handle->IsSameDocument())
+    return;
+
+  ClearEphemeralStorage();
```

Comment:
This is clearing storage on every load and we only want to clear it when the etld+1 changes

Reply:
My understanding of the requirements from @pes10k is that a main frame load even to the same etld+1 should clear the storage.

Reply:
the 3p storage is cleared even when `ClearEphemeralStorage` is commented out. If we use the tab helper to clear the storage then it's trivial to keep it for the lifetime of the eTLD which is better than keeping it just for the page load, but the code does not appear to be working as you expect because this method is not needed to achieve that behavior.

Reply:
I'm still debugging, but I believe this is happening because the StorageArea you are creating doesn't have a matching namespace in the browser process. I thought that caused it to crash, but that isn't happening. Instead what appears to be happening is that the StorageArea is recreated on every page load (making this call unnecessary)

Reply:
This isn't great, since we are relying on a side-effect of our misuse of the internal APIs. I think we are going to fix this so that the browser-side has a a properly matching namespace.